### PR TITLE
add tcp route

### PR DESCRIPTION
**Pull request by kl52752**
Add AppNet Tcp Route resource to rGraph

### CODE REVIEW

File: pkg/cloud/rgraph/rnode/backendservice/node.go [hunk 1]
```diff
@@ -109,6 +106,14 @@ func (n *backendServiceNode) Actions(got rnode.Node) ([]exec.Action, error) {
 
 	return nil, fmt.Errorf("BackendServiceNode: invalid plan op %s", op)
 }
+func (n *backendServiceNode) Actions(got rnode.Node) ([]exec.Action, error) {
```
[the remark this answers is not in the document]
Comment (kl52752) [reply pkg/cloud/rgraph/rnode/backendservice/node.go:109]: yes just for logging

File: pkg/cloud/rgraph/rnode/tcproute/builder.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+func NewBuilder(id *cloud.ResourceID) rnode.Builder {
+	b := &builder{}
+	b.Defaults(id)
+	b.resourceURLParser = cloud.ParseLegacyResourceURL
+	return b
+}
+
+// NewBuilderWithResource creates builder for tcp route
+// with predefined resource.
+func NewBuilderWithResource(r TcpRoute, parser urlParserFunc) rnode.Builder {
```
[the remark this answers is not in the document]
Comment (kl52752) [reply pkg/cloud/rgraph/rnode/tcproute/builder.go:46]: TcpRoute supports legacy urls I thought that in future they might support current one :)

File: pkg/cloud/rgraph/rnode/tcproute/node.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"google.golang.org/api/networkservices/v1"
+	beta "google.golang.org/api/networkservices/v1beta1"
+)
+
+type tcpRouteNode struct {
+	rnode.NodeBase
+	resource TcpRoute
+}
+
+var _ rnode.Node = (*tcpRouteNode)(nil)
```
[the remark this answers is not in the document]
Comment (kl52752) [reply pkg/cloud/rgraph/rnode/tcproute/node.go:33]: done

File: pkg/cloud/rgraph/rnode/tcproute/builder.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+// builder implements node.Builder.
+var _ rnode.Builder = (*builder)(nil)
+
+func (b *builder) Resource() rnode.UntypedResource { return b.resource }
+
+func (b *builder) SetResource(u rnode.UntypedResource) error {
+	r, ok := u.(TcpRoute)
+	if !ok {
+		return fmt.Errorf("cannot set TcpRoute from untyped resource")
```
[the remark this answers is not in the document]
Comment (kl52752) [reply pkg/cloud/rgraph/rnode/tcproute/builder.go:67]: done

File: pkg/cloud/utils.go [hunk 5]
```diff
@@ -142,6 +142,10 @@ func (r *ResourceID) String() string {
 // legacyApiGroupRegex is used to extract API Group from legacy path in format
 var apiGroupRegex = regexp.MustCompile(`([a-z]*)(\.googleapis\.com)?\/(alpha|beta|v1|v1alpha1|v1beta1)/projects`)
 
+// legacyApiGroupRegex is used to extract API Group from legacy path in format
+// https://<api_group>.googleapis.com/<ver>/projects/<resource_path>
+var legacyApiGroupRegex = regexp.MustCompile(`([a-z]*)\.googleapis\.com/(alpha|beta|v1|v1alpha1|v1beta1)/projects`)
```
Comment (bowei) [inline pkg/cloud/utils.go:147]: where is this used?

**Comment by kl52752** [comment]
> This mostly looks good -- but seems like we have to get the other changes in first.

yes, that's why this PR is marked as wip. It is dependent on PlaceholderType and ActionUpdate.

File: pkg/cloud/utils.go (hunk 5 above)
Comment (kl52752) [reply pkg/cloud/utils.go:147]: wrong rebase - removed

**Comment by kl52752** [comment]
> This mostly looks good -- but seems like we have to get the other changes in first.

PlaceholderType PR was merged, I applied those changes. It is now ready to review

**Comment by kl52752** [comment]
cc/ @mag-kol

**Pull request merged**